### PR TITLE
httpc: reset headers on redirect

### LINKED ISSUE #6101: Http client: Response contains incorrect headers in case of redirection
**Bug description**

* OS: all
* Architecture: all
* Tarantool version: all

Http client in case of redirection (3xx status code in case of `follow location`) returns response with incorrect headers 
(headers from response that contains 3xx status code) 

**Steps to reproduce**

Reproducer script:

```lua
#!/usr/bin/env tarantool
http = require('http.client')
http.request('GET', 'https://tarantool.io/release/2.7/installer.sh', nil, {verbose = true}).headers
```

Result:
```
* Hostname tarantool.io was found in DNS cache
*   Trying 188.93.56.54:443...
* Connected to tarantool.io (188.93.56.54) port 443 (#4)
* ALPN, offering http/1.1
* NPN, negotiated HTTP1.1
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: C=RU; L=\U041C\U043E\U0441\U043A\U0432\U0430; O=LLC Mail.Ru; CN=tarantool.io
*  start date: Nov 13 00:00:00 2020 GMT
*  expire date: Dec 14 23:59:59 2021 GMT
*  subjectAltName: host "tarantool.io" matched cert's "tarantool.io"
*  issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=GeoTrust RSA CA 2018
*  SSL certificate verify ok.
> GET /release/2.7/installer.sh HTTP/1.1
Host: tarantool.io
Accept: */*
Connection: close

* Mark bundle as not supporting multiuse
< HTTP/1.1 301 Moved Permanently
< Server: nginx/1.9.2
< Date: Wed, 26 May 2021 11:54:52 GMT
< Content-Type: text/html
< Content-Length: 184
< Connection: close
< Location: https://www.tarantool.io/release/2.7/installer.sh
< X-XSS-Protection: 0
< 
* Closing connection 4
* Issue another request to this URL: 'https://www.tarantool.io/release/2.7/installer.sh'
* Hostname www.tarantool.io was found in DNS cache
*   Trying 18.200.36.214:443...
* Connected to www.tarantool.io (18.200.36.214) port 443 (#5)
* ALPN, offering http/1.1
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=www.tarantool.io
*  start date: May 19 23:38:11 2021 GMT
*  expire date: Aug 17 23:38:11 2021 GMT
*  subjectAltName: host "www.tarantool.io" matched cert's "www.tarantool.io"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
> GET /release/2.7/installer.sh HTTP/1.1
Host: www.tarantool.io
Accept: */*
Connection: close

* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Connection: close
< Server: gunicorn/20.0.4
< Date: Wed, 26 May 2021 11:54:53 GMT
< Content-Type: text/plain; charset=utf-8
< Content-Disposition: attachment; filename="installer.sh"
< X-Frame-Options: DENY
< Content-Length: 10753
< Vary: Accept-Language, Cookie
< Content-Language: en
< X-Content-Type-Options: nosniff
< X-Xss-Protection: 1; mode=block
< Via: 1.1 vegur
< 
* Closing connection 5
---
- server: nginx/1.9.2
  content-type: text/html
  connection: close
  content-length: '184'
  x-xss-protection: '0'
  location: https://www.tarantool.io/release/2.7/installer.sh
  date: Wed, 26 May 2021 11:54:52 GMT
...
```

**Actual behavior**

After following location `response.headers` contains headers from  previous response:

**Expected behavior**

After redirection response must contains new headers

### PR DESCRIPTION
**Pull request by locker**
When a request is redirected, old headers are of no use anymore and should be dropped, but currently they are not. This results in invalid headers in the final response. Fix this.

Migrate the http client test to luatest and enable parallel test run for the app-luatest suite while we are at it.

Closes #6101

### CODE REVIEW

File: test/app-luatest/http_client_test.lua [hunk 1]
```diff
@@ -52,101 +50,129 @@ local function start_server(test, sock_family, sock_addr)
...
+        s:bind('127.0.0.1', 0)
+        sock_addr = string.format("%s:%d", s:name().host, s:name().port)
+        s:close()
+    else
+        assert(sock_family == 'AF_UNIX')
+        local path = os.tmpname()
+        os.remove(path)
+        local status = pcall(client.get, 'http://localhost/',
+                             {unix_socket = path})
+        t.skip_if(not status, "not supported")
```
Comment (Totktonada) [inline test/app-luatest/http_client_test.lua:71]: This is quite strange condition. If we'll broke unix sockets support in the http client, we'll not spot it in the test.

File: test/app-luatest/http_client_test.lua [hunk 2]
```diff
@@ -215,65 +242,69 @@ local function test_post_and_get(test, url, opts)
...
+    t.assert_not(status, "GET: exception on bad protocol - status")
+    t.assert_str_contains(json.encode(err), "Unsupported protocol",
+                          "GET: exception on bad protocol - error")
     status, err = pcall(http.post, http, "htp://mail.ru", "")
-    test:ok(not status and string.find(json.encode(err),
-                        "Unsupported protocol"),
-                        "POST: exception on bad protocol")
+    t.assert_not(status, "POST: exception on bad protocol - status")
+    t.assert_str_contains(json.encode(err), "Unsupported protocol",
+                          "POST: exception on bad protocol - error")
```
Comment (Totktonada) [inline test/app-luatest/http_client_test.lua:332]: Just `tostring()` works good for box errors.

File: test/app-luatest/http_client_test.lua [hunk 3]
```diff
@@ -215,65 +242,69 @@ local function test_post_and_get(test, url, opts)
...
 
     r = responses.bad_get
-    test:is(r.status, 404, "GET: http page not exists")
-    test:is(r.reason, 'Unknown', '404 - Unknown')
-    test:isnt(r.body:len(), 0, "GET: not empty body page not exists")
-    test:ok(string.find(r.body, "Not Found"),
-                "GET: right body page not exists")
+    t.assert_equals(r.status, 404, "GET: http page not exists")
+    t.assert_equals(r.reason, 'Unknown', '404 - Unknown')
+    t.assert(r.body:len() ~= 0, "GET: not empty body page not exists")
```
Comment (Totktonada) [inline test/app-luatest/http_client_test.lua:313]: It is better to use `t.assert_not_equals()`. In case of a failure, it'll show the bad value.

**Comment by Totktonada** [comment]
First two patches: ok. I'll continue.

**Review by Totktonada — APPROVED** [review]
LGTM.

**Review by kyukhin — APPROVED** [review]

**Comment by locker** [comment]
Cherry-picked to 1.10 and 2.8.

**Pull request merged**